gRFC: host rewrite literal (#1)
zhiyanfoo/grpc-proposal: A84-xds-authority-literal-rewrite.md

            +------------------+      target one of Svc A/B/C)
                     |
                     |    <- (gRPC xDS client reuses the same CDS target for all connections to
                     |        the proxy)
                 +---------+
                 |  Proxy  |
                 +---------+
                      |   <- (Proxy routing logic based on Authority header)
      .---------------+--------------.
     /                |               \
+--------+       +--------+       +--------+
| Svc A  |       | Svc B  |       | Svc C  |
+--------+       +--------+       +--------+
```

### Related Proposals:
* [gRFC A29: xDS-Based Security for gRPC Clients and Servers][A29]
* [gRFC A81: xDS Authority Rewriting][A81]
* [gRFC A86: xDS-Based HTTP CONNECT][A86]

## Proposal

Implement [`host_rewrite_literal`][route_action] feature in gRPC xDS client, enabling explicit
authority header rewrites. This will be conditioned on `trusted_xds_server` as described in
[gRFC A81][A81].

### xDS Resource Validation

When validating an RDS resource, if the `trusted_xds_server` server option is present for this

> xDS server in the bootstrap config, we will look at the `RouteAction.host_rewrite_literal`

Author: We can check to see precdence and see what to do if multiple host rewrite fields are set.

field. The string value of this field will be included in xds resource for routes.

### xDS ConfigSelector Changes

This value is propagated to the `ConfigSelector` in the xds resolver, and stored in it's struct
as part of the route configuration.

> ### xds_cluster_impl LB Policy Changes

Author: title of this is wrong, we aren't touching xds_cluster_impl. We modify the transport directly.

Title
----
* Author(s): Zhiyan Foo

> * Approver: a11r

Reviewer: TBD also.

* Status: Draft
* Implemented in: TBD
* Last updated: 2024-11-06
* Discussion at: TBD (filled after thread exists)

## Abstract

Propose extending gRPC's xDS client to honour `host_rewrite_literal` in route configuration for
authority header rewriting when `trusted_xds_server` is set.

## Background

> gRPC xDS currently supports `auto_host_rewrite` as per gRFC A81 but lacks `host_rewrite_literal`

Reviewer: ```suggestion
gRPC xDS currently supports `auto_host_rewrite` as per [gRFC A81][A81] but lacks `host_rewrite_literal`
```

support, useful when a single cluster targets a reverse proxy routing based on the authority
header.
```
            +------------------+
            |  gRPC xDS Client |  <- (RDS configuration sets `host_rewrite_literal` to
            +------------------+      target one of Svc A/B/C)
                     |
                     |    <- (gRPC xDS client reuses the same CDS target for all connections to
                     |        the proxy)
                 +---------+
                 |  Proxy  |
                 +---------+
                      |   <- (Proxy routing logic based on Authority header)
      .---------------+--------------.
     /                |               \
+--------+       +--------+       +--------+
| Svc A  |       | Svc B  |       | Svc C  |
+--------+       +--------+       +--------+
```

### Related Proposals:

> * [A81: `auto_host_rewrite` in gRPC xDS](https://github.com/grpc/proposal/blob/master/A81-xds-routes.md)
> * [A86: xDS-Based HTTP CONNECT](https://github.com/grpc/proposal/pull/455)

Reviewer: ```suggestion
* [gRFC A29: xDS-Based Security for gRPC Clients and Servers][A29]
* [gRFC A81: xDS Authority Rewriting][A81]
* [gRFC A86: xDS-Based HTTP CONNECT][A86]

[A29]: A29-xds-tls-security.md
[A81]: A81-xds-authority-rewriting.md
[A86]: A86-xds-http-connect.md
```

(make sure those links worked, I just showed you the format)

## Proposal

> Implement `host_rewrite_literal` feature in gRPC xDS client, enabling explicit authority header

Reviewer: A link to the envoy documentation for this field would be useful to readers.

> rewrites. This should be conditioned on `trusted_xds_server`.

Reviewer: ```suggestion
rewrites. This will be conditioned on `trusted_xds_server`.
```

I'd also add a link to gRFC A81 which describes this feature.

> 

Reviewer: You'll need to explain exactly which component of gRPC xDS are affected by this change: is setting the authority done in the config selector? Where is the authority currently set? Is there a chance that there are bad interops between the other places where we write an authority header? How will this interact with other settings such as `auto_host_rewrite`? We need to validate the config to make sure both are not set I think.

I would map those paragraphs the same way A81 does:
### xDS Resource Validation
### xDS ConfigSelector Changes
### xds_cluster_impl LB Policy Changes

Reviewer: This has good guidance on how/why this is important: https://github.com/grpc/proposal/pull/414#discussion_r1837279888

### Temporary environment variable protection

Feature guarded by `GRPC_XDS_EXPERIMENTAL_AUTHORIY_LITERAL_REWRITE`, disabled by default.

## Rationale

Alternative approaches considered.
- Exclusively use the existing `auto_host_rewrite` feature. This would require a CDS cluster per

>   upstream target, which mean that connections would not be re-used between gRPC xDS client and

Reviewer: ```suggestion
upstream target, which means that there would be one connection per service behind the proxy
between each gRPC clients and the proxy.
```

  the proxy.
- Using the HTTP CONNECT protocol has the same drawback in that connections to the proxy that
  target different upstream services won't reuse the same connection.
   - in the case where instead of a single proxy in between the target upstream service there
     are two proxies (e.g. client -> egress proxy -> ingress proxy), this would not only prevent
     connection reuse between the initial client to the proxy, but also between the intermediary
     proxies (e.g. between the egress and ingress proxy).

Even if connection pooling was not a major concern, this addition will bring gRPC xDS client
functionality closer to parity with Envoy's capabilities, simplifying configurations for users
migrating to or using both systems concurrently.

## Implementation

> We can contribute an implementation of this to golang. From our initial drafts, the changes
> required to support this are minimal.

Reviewer: ```suggestion
We will contribute an implementation to Go.
```

I would wait for some feedback on the gRFC before investing too much in the implementation (tests, etc.). Happy to do a first review of the PR though.
